packages/ui-predicate-react/src/ui-predicate.js

```diff
@@ -0,0 +1,166 @@
+    this.setState({ isInAddCompoundMode });
+    if (this.props.onAddCompoundModeChange)
+      this.props.onAddCompoundModeChange(isInAddCompoundMode);
+  }
+
+  triggerChanged() {
+    // update the state with the new root
+    this.setState({
+      root: this.state.ctrl.root,
+    });
+    // callback onChanged when some predicates where changed
+    if (this.props.onChange) this.props.onChange(this.state.ctrl.toJSON());
+  }
+
+  add(predicate) {
+    return this.state.ctrl.add({
+      where: predicate,
+      how: 'after',
+      type: this.state.isInAddCompoundMode
+        ? 'CompoundPredicate'
+        : 'ComparisonPredicate',
+    });
+  }
+  remove(predicate) {
+    return this.state.ctrl.remove(predicate);
+  }
+  setPredicateLogicalType_id(predicate, logicalType_id) {
+    return this.state.ctrl.setPredicateLogicalType_id(
+      predicate,
+      logicalType_id
+    );
+  }
+  setPredicateTarget_id(predicate, target_id) {
+    return this.state.ctrl.setPredicateTarget_id(predicate, target_id);
+  }
+  setPredicateOperator_id(predicate, operator_id) {
+    return this.state.ctrl.setPredicateOperator_id(predicate, operator_id);
+  }
+  getArgumentTypeComponentById(argumentType_id) {
+    return this.state.ctrl.getArgumentTypeComponentById(argumentType_id);
+  }
+  setArgumentValue(predicate, value) {
+    return this.state.ctrl.setArgumentValue(predicate, value);
+  }
+  getUIComponent(name) {
+    return this.state.ctrl.getUIComponent(name);
+  }
+
+  componentDidMount() {
+    window.addEventListener('keyup', this.onAltReleased);
+    window.addEventListener('keydown', this.onAltPressed);
+
+    UIPredicateCoreReact({
+      data: this.props.data,
+      columns: this.props.columns,
+      ui: this.props.ui,
+    }).then(
+      ctrl => {
+        // Listen for changes
+        ctrl.on('changed', this.triggerChanged);
+        // Set current state to indicate we're ready
+        this.setState(
+          {
+            ctrl,
+            root: ctrl.root,
+            isCoreReady: true,
+          },
+          this.handleInitialized
+        );
+      },
+      err => {
+        const initializationFailedError = Object.assign(
+          new InitializationFailed(),
+          { cause: err }
+        );
+        throw initializationFailedError;
+      }
+    );
+  }
+  componentWillUnmount() {
+    if (this.state.ctrl) this.state.ctrl.off();
+    window.removeEventListener('keyup', this.onAltReleased);
+    window.removeEventListener('keydown', this.onAltPressed);
+  }
+
+  render() {
+    const { isCoreReady, root } = this.state;
+    const { columns, className } = this.props;
+    return (
+      <div className={`ui-predicate__main ${className}`}>
+        {!isCoreReady && Boolean(this.props.loader) && this.props.loader}
+        {Boolean(isCoreReady) && (
+          <UIPredicateCompound
+            predicate={root}
+            columns={columns}
+            getAddCompoundMode={this.getAddCompoundMode}
```

Comment:
UIPredicateCompound should only have `predicate` & `columns` props. 
Why does it have add, remove, setPredicateLogicalType_id etc.. here?

Reply:
I will have to look into it again to be sure. In the Vue version, the properties are injected. In React we need to pass them through the hierarchy manually.

Reply:
So yeah... These properties are passed along into the components manually. 
I could rewrite the code a bit to make it less verbose (picking the function from `this.state.ctrl` and then destructure them into the component). Or... I could look into how `context` works in React. I read that it works in similar ways as `inject` in *Vue*. But haven't worked with `context` before (still new-ish to React).
What do you think?